zk/dnshostprovider.go

```diff
@@ -30,33 +74,37 @@ func (hp *DNSHostProvider) Init(servers []string) error {
 		lookupHost = net.LookupHost
 	}
 
-	found := []string{}
-	for _, server := range servers {
+	if len(hp.unresolvedServers) == 0 {
+		return true, nil
+	}
+
+	found := make([]string, 0, len(hp.unresolvedServers))
+	for server := range hp.unresolvedServers {
 		host, port, err := net.SplitHostPort(server)
 		if err != nil {
-			return err
+			return false, err
 		}
 		addrs, err := lookupHost(host)
```

Comment:
one more suggestion: lookupHost may take a while. you might not want to hold the lock while it's executing. consider releasing it, and re-locking once it returns (if there's more work to do)

Reply:
i guess you could just hold off on obtaining the lock until you start manipulating shared state (starting with `hp.servers = append ...`)

Reply:
I think that's a very good suggestion @jdef .  However this optimization makes `unresolvedServers`, `sleep`, and `lookupHost` not thread safe. It is actually acceptable, since they are only used during lookup. 

I updated the commit based on your suggestion, and also the mutex now only protects `servers`, `curr`, and `last`.

Also updated the test to use channel instead to simulate a server that gets back online later. Previously I changed `lookupHost` inflight which would be a race in the new implementation.